sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package expansionx contains utilities for starting expansion services for
+// cross-language transforms.
+package expansionx
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+)
+
+// URL is a type used to differentiate between strings and URLs
+type URL string
+
+const (
+	apacheRepository = URL("https://repo.maven.apache.org/maven2")
+	beamGroupID      = "org/apache/beam"
+	jarCache         = "~/.apache_beam/cache/jars"
+)
+
+// GetBeamJar checks a temporary directory for the desired Beam JAR, downloads the
+// appropriate JAR from Maven if not present, then returns the file path to the
+// JAR.
+func GetBeamJar(gradleTarget, version string) (string, error) {
+	strippedTarget := dropEndOfGradleTarget(gradleTarget)
+	fullURL, jarName := getURLForBeamJar(strippedTarget, version)
+
+	resp, err := http.Get(string(fullURL))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("Received non 200 response code, got %v", resp.StatusCode)
+	}
+
+	cacheDir := getCacheDir()
+	err = checkDir(cacheDir)
+	if err != nil {
+		return "", err
+	}
+
+	jarPath := filepath.Join(cacheDir, jarName)
+
+	if jarExists(jarPath) {
+		return jarPath, nil
+	}
```

Comment:
This should be the first part of the function, before the HTTP request. Part of the benefit of the cache is that one should be able to avoid making any HTTP requests if the cached jar is present.

This also lets one unit test the cache functionality in the  "cache hit" case, and not worry about the HTTP call. (Hurray coverage!)

Now ideally, we'd be doing proper validation, checksum/hash validation, but this is a simple, trusting cache.  (eg. If the jar is present, we should in principle be able to look up the checksum/hash with a different request to maven, and then if that also fails, fall back to re-downloading it, but that's not a v0 task.)

Reply:
100% correct, that's what I get for prototyping the requests first and the cache after the fact. 

FWIW the Python SDK also doesn't do the checksum/hash validation (there's a comment indicating that it should be added later)